File: sdks/typescript/src/enrichment.ts
import type { WorkEvent } from './protocol.js';

/**
 * Per-execution helper that injects `job_key`, `runner_id`, and (when set)
 * `runner_tags` into every event's `fields` map. Explicit caller values win
 * — the enricher uses `TryAdd` semantics, so a handler that puts an
 * `runner_id` in its own fields keeps that override.
 *
 * Mirrors `LogEnrichment` from the .NET SDK.
 */
export class LogEnrichment {
  readonly #jobKey: string;
  readonly #runnerId: string;
  readonly #serializedTags: string | undefined;

  constructor(jobKey: string, runnerId: string, runnerTags: readonly string[]) {
    this.#jobKey = jobKey;
    this.#runnerId = runnerId;
    this.#serializedTags = runnerTags.length === 0 ? undefined : JSON.stringify(runnerTags);
  }

  enrich(source: WorkEvent): WorkEvent {
    const fields: Record<string, string> = { ...(source.fields ?? {}) };
    if (!('job_key' in fields)) fields.job_key = this.#jobKey;
    if (!('runner_id' in fields)) fields.runner_id = this.#runnerId;
    if (this.#serializedTags !== undefined && !('runner_tags' in fields)) {
      fields.runner_tags = this.#serializedTags;
    }
    return { ...source, fields };
  }
}

File: sdks/go/enrichment.go
package croniq

import "encoding/json"

// serializeTags JSON-encodes the runner's tag list for inclusion as a
// single `runner_tags` log field. Returns "" when the runner has no tags.
func serializeTags(tags []string) string {
	if len(tags) == 0 {
		return ""
	}
	buf, err := json.Marshal(tags)
	if err != nil {
		return ""
	}
	return string(buf)
}

// enrichEvent auto-injects `job_key`, `runner_id`, and `runner_tags`
// into a log event's fields without overwriting caller-provided values.
// Returns a new event; the input is not mutated.
func enrichEvent(in WorkEvent, jobKey, runnerID, serializedTags string) WorkEvent {
	fields := make(map[string]string, len(in.Fields)+3)
	for k, v := range in.Fields {
		fields[k] = v
	}
	if _, ok := fields["job_key"]; !ok {
		fields["job_key"] = jobKey
	}
	if _, ok := fields["runner_id"]; !ok {
		fields["runner_id"] = runnerID
	}
	if serializedTags != "" {
		if _, ok := fields["runner_tags"]; !ok {
			fields["runner_tags"] = serializedTags
		}
	}
	return WorkEvent{
		Level:   in.Level,
		Message: in.Message,
		Fields:  fields,
	}
}
